src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.Reflection.cs

```diff
@@ -0,0 +1,219 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System;
+using System.Reflection;
+using System.Diagnostics.CodeAnalysis;
+
+namespace Microsoft.Extensions.DependencyInjection
+{
+    public static partial class ActivatorUtilities
+    {
+#if NET8_0_OR_GREATER // Use the faster ConstructorInvoker which also has alloc-free APIs when <= 4 parameters.
+        private static object ReflectionFactory_ServiceOnly_Fixed(
+            ConstructorInvoker invoker,
+            Type[] parameterTypes,
+            Type declaringType,
+            IServiceProvider serviceProvider)
+        {
+            Debug.Assert(parameterTypes.Length >= 1 && parameterTypes.Length <= FixedArgumentThreshold);
+            Debug.Assert(FixedArgumentThreshold == 4);
+
+            if (serviceProvider is null)
+                ThrowHelper_ArgumentNullException_serviceProvider();
+
+            object? arg1 = null;
+            object? arg2 = null;
+            object? arg3 = null;
+            object? arg4 = null;
+
+            switch (parameterTypes.Length)
+            {
+                case 4:
+                    arg4 = GetService(serviceProvider, parameterTypes[3], declaringType, false);
+                    goto case 3;
+                case 3:
+                    arg3 = GetService(serviceProvider, parameterTypes[2], declaringType, false);
+                    goto case 2;
+                case 2:
+                    arg2 = GetService(serviceProvider, parameterTypes[1], declaringType, false);
+                    goto case 1;
+                case 1:
+                    arg1 = GetService(serviceProvider, parameterTypes[0], declaringType, false);
+                    break;
+            }
+
+            return invoker.Invoke(arg1, arg2, arg3, arg4)!;
+        }
+
+        private static object ReflectionFactory_ServiceOnly_Span(
+            ConstructorInvoker invoker,
+            Type[] parameterTypes,
+            Type declaringType,
+            IServiceProvider serviceProvider)
+        {
+            ArgumentNullException.ThrowIfNull(nameof(serviceProvider));
+
+            object?[] arguments = new object?[parameterTypes.Length];
```

Comment:
I doubt it's worthwhile, but just asking... should we assume `stackalloc` here since we doubtfully have a huge number of constructor parameters? If we do that, do we want to fall-back to normal heap allocation like many string methods do?

Reply:
The only way to do a stackalloc for `object` is to declare a type that has `object1`, `object2` etc. (or the newer `[InlineArray(someLiteralNumber)]`)  but that means the length has to be known at compile-time. The general pattern here is for<=4 args leverage the new invoke APIs (no alloc) while >4 do a normal array alloc (just like the previous code). So bridging >5 and < some other higher threshold with helper classes is possible however there are diminishing value with this since >=5 ctor arguments is probably not going to be very common.